src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+                    break;
+
+                case BoundKind.SuppressNullableWarningExpression:
+                    VisitLvalue(((BoundSuppressNullableWarningExpression)node).Expression);
+                    break;
+
+                default:
+                    VisitRvalue(node);
+                    break;
             }
 
-            if (catchBlock.ExceptionFilterOpt != null)
+            if (_trackRegions && node == this.lastInRegion && this.regionPlace == RegionPlace.Inside) LeaveRegion();
+        }
+
+        protected virtual void VisitLvalue(BoundLocal node)
+        {
+            // no need for it to be previously assigned: it is on the left.
+        }
+
+        /// <summary>
+        /// Visit a boolean condition expression.
+        /// </summary>
+        /// <param name="node"></param>
+        protected void VisitCondition(BoundExpression node)
+        {
+            Visit(node);
+            AdjustConditionalState(node);
+        }
+
+        private void AdjustConditionalState(BoundExpression node)
+        {
+            if (IsConstantTrue(node))
             {
-                VisitCondition(catchBlock.ExceptionFilterOpt);
-                SetState(StateWhenTrue);
+                Unsplit();
+                SetConditionalState(this.State, UnreachableState());
+            }
+            else if (IsConstantFalse(node))
+            {
+                Unsplit();
+                SetConditionalState(UnreachableState(), this.State);
+            }
+            else if ((object)node.Type == null || node.Type.SpecialType != SpecialType.System_Boolean)
+            {
+                // a dynamic type or a type with operator true/false
+                Unsplit();
             }
 
-            VisitStatement(catchBlock.Body);
+            Split();
         }
 
-        protected virtual void VisitFinallyBlock(BoundStatement finallyBlock, ref TLocalState unsetInFinally)
+        /// <summary>
+        /// Visit a general expression, where we will only need to determine if variables are
+        /// assigned (or not). That is, we will not be needing AssignedWhenTrue and
+        /// AssignedWhenFalse.
+        /// </summary>
+        /// <param name="node"></param>
+        protected BoundNode VisitRvalue(BoundExpression node)
         {
-            VisitStatement(finallyBlock); // this should generate no pending branches
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
+            var result = Visit(node);
+            Unsplit();
+            return result;
         }
 
-        #endregion TryStatements
+        /// <summary>
+        /// Visit a statement.
+        /// </summary>
+        [DebuggerHidden]
+        protected virtual void VisitStatement(BoundStatement statement)
+        {
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
```

Comment:
this check seems to be duplicated in a few places.  might be better to have a clearer name and have that be asserted here.